File: android/app/src/main/java/com/aivisiondrone/groundstation/comms/Protocol.kt
package com.aivisiondrone.groundstation.comms

import org.json.JSONObject

/** Mirrors companion/comms/protocol.py's MessageType exactly - keep both in sync. */
object MessageType {
    const val TARGET_SELECT = "target_select"
    const val MODE_COMMAND = "mode_command"
    const val ABORT = "abort"
    const val ARM_COMMAND = "arm_command"
    const val ARM_COMMAND_RESULT = "arm_command_result"
    const val SET_FLIGHT_MODE = "set_flight_mode"
    const val RECORD_COMMAND = "record_command"
    const val RECORDING_STATE = "recording_state"
    const val LAND_CONFIRMATION_REQUEST = "land_confirmation_request"
    const val LAND_CONFIRMATION_RESPONSE = "land_confirmation_response"
    const val TRACKING_UPDATE = "tracking_update"
    const val DETECTIONS_UPDATE = "detections_update"
    const val GRID_SEARCH_UPDATE = "grid_search_update"
    const val TELEMETRY = "telemetry"
    const val HEALTH = "health"
    const val ACK = "ack"
    const val ERROR = "error"
    const val WEBRTC_OFFER = "webrtc_offer"
    const val WEBRTC_ANSWER = "webrtc_answer"
    const val TEACH_OBJECT = "teach_object"
    const val TEACH_RESULT = "teach_result"
    const val PING = "ping"
    const val PONG = "pong"
}

data class Envelope(
    val type: String,
    val seq: Int,
    val ts: Double,
    val payload: JSONObject,
) {
    fun toJson(): String =
        JSONObject().apply {
            put("type", type)
            put("seq", seq)
            put("ts", ts)
            put("payload", payload)
        }.toString()

    companion object {
        fun fromJson(raw: String): Envelope {
            val obj = JSONObject(raw)
            return Envelope(
                type = obj.getString("type"),
                seq = obj.getInt("seq"),
                ts = obj.getDouble("ts"),
                payload = obj.optJSONObject("payload") ?: JSONObject(),
            )
        }
    }
}

private var sequenceCounter = 0

fun makeEnvelope(type: String, payload: JSONObject): Envelope {
    sequenceCounter += 1
    return Envelope(
        type = type,
        seq = sequenceCounter,
        ts = System.currentTimeMillis() / 1000.0,
        payload = payload,
    )
}

File: companion/comms/protocol.py
from __future__ import annotations

import json
import time
from dataclasses import asdict, dataclass
from typing import Any


class MessageType:
    TARGET_SELECT = "target_select"
    MODE_COMMAND = "mode_command"
    ABORT = "abort"
    ARM_COMMAND = "arm_command"
    SET_FLIGHT_MODE = "set_flight_mode"
    RECORD_COMMAND = "record_command"
    RECORDING_STATE = "recording_state"
    LAND_CONFIRMATION_REQUEST = "land_confirmation_request"
    LAND_CONFIRMATION_RESPONSE = "land_confirmation_response"
    TRACKING_UPDATE = "tracking_update"
    DETECTIONS_UPDATE = "detections_update"
    TELEMETRY = "telemetry"
    HEALTH = "health"
    ACK = "ack"
    ERROR = "error"
    WEBRTC_OFFER = "webrtc_offer"
    WEBRTC_ANSWER = "webrtc_answer"


@dataclass
class Envelope:
    type: str
    seq: int
    ts: float
    payload: dict[str, Any]

    def to_json(self) -> str:
        return json.dumps(asdict(self))

    @classmethod
    def from_json(cls, raw: str) -> "Envelope":
        data = json.loads(raw)
        return cls(
            type=data["type"], seq=data["seq"], ts=data["ts"], payload=data.get("payload", {})
        )


class SequenceCounter:
    def __init__(self) -> None:
        self._seq = 0

    def next(self) -> int:
        self._seq += 1
        return self._seq


def make_envelope(msg_type: str, payload: dict[str, Any], seq: int) -> Envelope:
    return Envelope(type=msg_type, seq=seq, ts=time.time(), payload=payload)
